ksql-parser/src/main/java/io/confluent/ksql/parser/KsqlParser.java

```diff
@@ -16,33 +16,36 @@
 
 import io.confluent.ksql.metastore.MetaStore;
 import io.confluent.ksql.parser.SqlBaseParser.SingleStatementContext;
-import io.confluent.ksql.parser.exception.ParseFailedException;
-import io.confluent.ksql.parser.rewrite.StatementRewriteForStruct;
-import io.confluent.ksql.parser.tree.Node;
 import io.confluent.ksql.parser.tree.Statement;
-import io.confluent.ksql.util.DataSourceExtractor;
 import java.util.List;
 import java.util.Objects;
-import java.util.function.Consumer;
-import java.util.function.Function;
-import java.util.function.Predicate;
-import java.util.stream.Collectors;
-import org.antlr.v4.runtime.BaseErrorListener;
-import org.antlr.v4.runtime.CharStream;
-import org.antlr.v4.runtime.CharStreams;
-import org.antlr.v4.runtime.CommonTokenStream;
-import org.antlr.v4.runtime.ParserRuleContext;
-import org.antlr.v4.runtime.RecognitionException;
-import org.antlr.v4.runtime.Recognizer;
-import org.antlr.v4.runtime.atn.PredictionMode;
-import org.antlr.v4.runtime.misc.Interval;
-import org.antlr.v4.runtime.misc.ParseCancellationException;
-
-// CHECKSTYLE_RULES.OFF: ClassDataAbstractionCoupling
-public class KsqlParser {
-  // CHECKSTYLE_RULES.ON: ClassDataAbstractionCoupling
-
-  public static final class ParsedStatement {
+
+/**
+ * A SQL parser.
+ */
+public interface KsqlParser {
+
+  /**
+   * Parse the supplied {@code sql} into a list of statements.
+   *
+   * @param sql the sql to parse.
+   * @return the list of parsed statements.
+   */
+  List<ParsedStatement> parse(String sql);
+
+  /**
+   * Prepare the supplied {@code statement}.
+   *
+   * <p>The abstract syntax tree is created using the preexisting entities within the
+   * {@code metastore}.
+   *
+   * @param statement the statement to build
+   * @param metaStore the meta store of existing entities.
+   * @return the prepared statement.
+   */
+  PreparedStatement<?> prepare(ParsedStatement statement, MetaStore metaStore);
```

Comment:
Why returning PreparedStatement<?> instead of PreparedStatement<? extends Statement>? 

For readability, the 2nd choice is easier to read when wanted to know what type of statement is allowed by the PreparedStatement without going to the class to see the T type. The Java compiler will also throw compilation errors if a developer attempts to return something different a future bug fix.

BTW, I've seen this pattern to often in the code. I also was wondering if there is a benefit on returning <?> instead of specifying the correct type.

Reply:
> Why returning PreparedStatement<?> instead of PreparedStatement<? extends Statement>?

The two are the same. The generic parameter of `PreparedStatement` is already bound to `extend Statement`.

> BTW, I've seen this pattern to often in the code. I also was wondering if there is a benefit on returning <?> instead of specifying the correct type.

There is no way of knowing from the `statement` what the correct type would be.  Admittedly, now that it's only preparing a single statement we could make the method itself generic and internally do the cast to a specific type:

```java
<TT extends Statement> PreparedStatement<TT> prepare(ParsedStatement statement, ...)
``` 

I'll see if that would simplify code anywhere...

Reply:
Thanks for the explanation. I agree using any option. I was just wondering about the option 1 because I haven't seen code using <?> too often. But it's good for me.